Add checksum file of released binaries

twang2218 opened the pull request:
Add 'sha256sum.txt' and 'md5sum.txt' for release binaries.
Also fixed a minor bug which `bin/docker-machine` binary shouldn't be shown on the released file list.

Signed-off-by: Tao Wang <twang2218@gmail.com>

nathanleclaire commented:
> Also fixed a minor bug which bin/docker-machine binary shouldn't be shown on the released file list.

Ah nice, yeah that's been bugging me for a while.

nathanleclaire commented:
Any particular motivation to upload the checksums as files as well as print them in the description body?  I guess in case you need to do verification in some sort of automated system?

twang2218 commented:
Yes, I have a script to download latest released docker softwares to my local server. To be able to verify the downloaded file, I need the script to get the checksum of the files.

Currently, I have to manually copy/paste the checksum from the release notes to my script. But I want to make it automatically. I tried to parse the release notes to get the checksum, however, it's a little bit hacky. So, I think the better way is to have the checksum files in the first place.

nathanleclaire commented on script/release.sh:
```diff
@@ -139,13 +139,20 @@ CONTRIBUTORS=$(git log "${LAST_RELEASE_VERSION}".. --format="%aN" --reverse | so
 done
 
+# Generate Checksum file
+(
```
Why not do this in the top level directory instead of `cd`-in?

twang2218 replied:
`cd` into the `bin` directory will make the result more clean:

With `cd bin`, the generated `sha256sum.txt` is:

```
efc5306920194c3ac4be3baaa09b4cbb398db0e6e6afe0e241287b3c4db9dcfe  docker-machine-Darwin-x86_64
87bc42fa15e1fdc1d543256882828eb6bc7728f2cc66246c2aa33bef3efdb41a  docker-machine-Linux-armhf
74294a6593310bb273c99a915bf95fa5990a995dbbbc9f6cd0b26cfb640ce43c  docker-machine-Linux-x86_64
f91971878808192da79ffc6cf25fb5aa43df9441d8fb4c40979552e81c5a8bf2  docker-machine-Windows-i386.exe
c8a07bc01348bb69810f63a5f39d2dbfa0376037dbcaa9d4f45431419254406a  docker-machine-Windows-x86_64.exe
```

Without `cd bin`, the result is:

```bash
f80862f0f0fe77f756ad49794b61be5cf14061095e4233f65c23ea4ca63a0c29  bin/docker-machine-Darwin-x86_64
b219f94aea6bac213a257c4e14df9609b4766646573193b67e1778feacc08742  bin/docker-machine-Linux-armhf
b77ddd4d7f47902c6082375962caa454e9466e2e484ed6cc1dea08da50d4e285  bin/docker-machine-Linux-x86_64
83092b781f8b818cee23ad2f93949de88689b986aa3ed15d060d5c069fef1b15  bin/docker-machine-Windows-i386.exe
ac5180b920ad97e1dd18a03cf221f84b42bb0038e0a1871de8bbefd6d2a849ea  bin/docker-machine-Windows-x86_64.exe
```

There will be a `bin/` path prefixed on the filename in the checksum file by default. Although the prefix can be removed later, it has to be done on both checksum files, so `cd bin` may be a cleaner way.

nathanleclaire commented on script/release.sh:
```diff
@@ -139,13 +139,20 @@ CONTRIBUTORS=$(git log "${LAST_RELEASE_VERSION}".. --format="%aN" --reverse | so
+# Generate Checksum file
+(
+    cd bin
+    shasum -a 256 docker-machine-* > ../sha256sum.txt
```
Can you please use `openssl dgst` here instead?  It should be present on both OSX and Linux and `md5sum` is not present on my OSX box.

twang2218 replied:
If replace the `shasum -a 256` with `openssl dgst -sha256` the checksum format will be different.

```bash
shasum -a 256 docker-machine-* > ../sha256sum.txt
```

will generate result

```bash
3fb225ece68c6517482a6109e737f77dda000ee4d35685e1977f7ed0f4587ef7  docker-machine-Darwin-x86_64
c844d3d37a677ab660fd9f14d7fd0fe1b323ae12ee33d8cbe73907a274d530ab  docker-machine-Linux-armhf
c6e35da73d81c5fc0236e786696397b414a41df8580cf36f8f1f27527222daa9  docker-machine-Linux-x86_64
2d1725b06bafc77ba14a2794696d03f2d29c85984d3211069f2d2ee2c6278a70  docker-machine-Windows-i386.exe
91ec3ad748740d125efc15ce08a454d0d9c03ecf465ee3ec3ff50a98716cc33c  docker-machine-Windows-x86_64.exe
```

And with

```bash
openssl dgst -sha256 docker-machine-* > ../sha256sum.txt
```

will generate the checksum in following format:

```bash
SHA256(docker-machine-Darwin-x86_64)= 3fb225ece68c6517482a6109e737f77dda000ee4d35685e1977f7ed0f4587ef7
SHA256(docker-machine-Linux-armhf)= c844d3d37a677ab660fd9f14d7fd0fe1b323ae12ee33d8cbe73907a274d530ab
SHA256(docker-machine-Linux-x86_64)= c6e35da73d81c5fc0236e786696397b414a41df8580cf36f8f1f27527222daa9
SHA256(docker-machine-Windows-i386.exe)= 2d1725b06bafc77ba14a2794696d03f2d29c85984d3211069f2d2ee2c6278a70
SHA256(docker-machine-Windows-x86_64.exe)= 91ec3ad748740d125efc15ce08a454d0d9c03ecf465ee3ec3ff50a98716cc33c
```

And `openssl dgst` cannot verify the checksum file it generated, unless it's been signed. So, I think it's better keeping `shasum` and `md5sum` format for checksum files, so user can verify the downloaded files simply by `shasum -a 256 -c sha256sum.txt`.

And to keep the same format of `md5sum` and `shasum` generated checksum file, I think it's better to use the `shasum` and `md5sum` for the job. I installed `md5sum` on my macOS via `brew install md5sha1sum`.

Do you think it is necessary to remove the `md5sum` dependency? If so, I think I might be able to construct the checksum file from the checksum calculated in the above section, such as:

```bash
CHECKSUM=""
rm sha256sum.txt md5sum.txt
for file in $(ls bin/docker-machine-*); do
  SHA256=$(openssl dgst -sha256 < "${file}")
  MD5=$(openssl dgst -md5 < "${file}")
  LINE=$(printf "\n * **%s**\n  * sha256 \`%s\`\n  * md5 \`%s\`\n\n" "$(basename ${file})" "${SHA256}" "${MD5}")
  CHECKSUM="${CHECKSUM}${LINE}"
  echo "${SHA256}  ${file:4}" >> sha256sum.txt
  echo "${MD5}  ${file:4}" >> md5sum.txt
done
```

What do you think?

nathanleclaire commented on script/release.sh:
```diff
@@ -229,6 +236,21 @@ for file in $(ls bin/docker-machine-*); do
 done
 
+display "Uploading checksums"
+for file in sha256sum.txt md5sum.txt; do
```
Nitpick, let's please call it `sha256sums.txt` to indicate it contains multiple checksums

twang2218 replied:
Ok, I will change the display title to `Uploading sha256sums.txt and md5sums.txt`

nathanleclaire commented:
Cool, I'm game for this change (and thanks for it @twang2218), just have some comments I'd like to get addressed if I'm going to merge it.

twang2218 commented:
I removed the `md5sum` dependency by constructing the checksum files from the already calculated checksums. I think this might be cleaner.

And I also fixed the display title for uploading checksums.

nathanleclaire commented:
Cool.  Thanks for the PR -- LGTM
